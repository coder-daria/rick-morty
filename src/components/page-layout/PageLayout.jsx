import React, { useCallback, useState } from 'react';
import { instanceOf, func, number, shape, bool } from 'prop-types';
import { Button } from 'antd';

import { EMPTY_ARRAY, EMPTY_OBJECT } from '../../common/constants';
import { selectedListItemVar, isDrawerOpenVar } from '../../apollo/cache';

import { StyledTable, Bold, PageInfo } from './PageLayout.styles';

const getActionColumnData = () => ({
  title: 'Action',
  key: 'action',
  render: data => (
    <Button
      onClick={() => {
        selectedListItemVar(data.id);
        isDrawerOpenVar(true);
      }}
    >
      View
    </Button>
  ),
});

function PageLayout({ columns, loading, pageInfo, tableData, setCurrentPage }) {
  const [nextPage, setNextPage] = useState(2);

  const handleSetCurrentPage = useCallback(
    e => {
      if (nextPage === e.current) {
        setCurrentPage(e.current);
        setNextPage(e.current + 1);
      }
    },
    [nextPage, setCurrentPage],
  );

  return (
    <div>
      <PageInfo>
        <span>
          Showing<Bold>{tableData?.length}</Bold>
          of<Bold>{pageInfo?.count}</Bold> entries
        </span>
      </PageInfo>
      <StyledTable
        rowKey={record => record.id}
        tableLayout='fixed'
        loading={loading}
        onChange={handleSetCurrentPage}
        columns={[...columns, getActionColumnData()]}
        dataSource={tableData}
        pagination={{
          position: ['topLeft'],
          pageSize: 20,
          total: pageInfo.count,
          simple: true,
        }}
        scroll={{ y: 700 }}
      />
    </div>
  );
}

PageLayout.propTypes = {
  loading: bool.isRequired,
  columns: instanceOf(Array),
  tableData: instanceOf(Array),
  pageInfo: shape({
    pages: number,
    count: number,
    next: number,
    prev: number,
  }),
  setCurrentPage: func.isRequired,
};

PageLayout.defaultProps = {
  columns: EMPTY_ARRAY,
  tableData: EMPTY_ARRAY,
  pageInfo: EMPTY_OBJECT,
};

export default PageLayout;
