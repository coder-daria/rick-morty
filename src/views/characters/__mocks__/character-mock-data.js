export const LOCATION_NAME = 'Earth';

export const CHARACTER_NOT_FORMATED_MOCKED_DETAILS = {
  location: {
    name: LOCATION_NAME,
  },
  origin: {
    name: LOCATION_NAME,
  },
};

export const CHARACTER_FORMATED_MOCKED_DETAILS = {
  location: LOCATION_NAME,
  origin: LOCATION_NAME,
};

export const CHARACTER_DETAILED_MOCKED_DATA = {
  created: '11/25/2020, 4:59:08 PM',
  episode: [],
  gender: 'Male',
  image: '',
  name: 'Rick Sanchez',
  species: 'Alien',
  status: 'Alive',
  type: '',
};
