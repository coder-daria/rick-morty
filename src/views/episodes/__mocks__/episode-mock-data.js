export const EPISODE_NOT_FORMATED_MOCKED_DETAILS = {
  air_date: '11/25/2020, 4:59:08 PM',
};

export const EPISODE_FORMATED_MOCKED_DETAILS = {
  'air date': '11/25/2020, 4:59:08 PM',
};

export const EPISODE_DETAILS_MOCKED_DATA = {
  name: 'Sunny day',
  episode: '0E12',
  created: '11/25/2019, 4:59:08 PM',
  characters: [{ name: 'Tom' }],
};
