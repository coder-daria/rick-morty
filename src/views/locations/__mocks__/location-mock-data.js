export const LOCATION_DETAILS_MOCKED_DATA = {
  name: 'Earth',
  created: '11/25/2019',
  dimension: '',
  type: '',
  residents: [{ name: 'Tom' }],
};
