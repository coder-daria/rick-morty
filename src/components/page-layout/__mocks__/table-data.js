import { CHARACTER_NOT_FORMATED_MOCKED_DETAILS } from '../../../views/characters/__mocks__/character-mock-data';

export const PAGE_INFO = {
  pages: 1,
  count: 1,
  next: null,
  prev: null,
};

export const TABLE_DATA = {
  columns: [
    {
      title: 'Name',
      dataIndex: 'name',
      key: 'name',
    },
    {
      title: 'Species',
      dataIndex: 'species',
      key: 'species',
    },
    {
      title: 'Origin',
      dataIndex: 'origin',
      key: 'origin',
    },
    {
      title: 'Location',
      dataIndex: 'location',
      key: 'location',
    },
  ],
  loading: false,
  pageInfo: PAGE_INFO,
  tableData: [
    {
      id: 1,
      name: 'Rick Sanchez',
      species: 'Alien',
      ...CHARACTER_NOT_FORMATED_MOCKED_DETAILS,
    },
  ],
  toggleDrawer: jest.fn(),
  setCurrentPage: jest.fn(),
};
