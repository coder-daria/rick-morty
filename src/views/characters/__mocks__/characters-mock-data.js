import { PAGE_INFO } from '../../../components/page-layout/__mocks__/table-data';
import { CHARACTER_NOT_FORMATED_MOCKED_DETAILS } from './character-mock-data';

const CHARACTER_TABLE_MOCKED_DATA = {
  id: 1,
  name: 'Rick Sanchez',
  species: 'Alien',
  ...CHARACTER_NOT_FORMATED_MOCKED_DETAILS,
};

export const CHARACTERS_MOCKED_DATA = {
  characters: {
    results: [CHARACTER_TABLE_MOCKED_DATA],
    info: PAGE_INFO,
  },
};
