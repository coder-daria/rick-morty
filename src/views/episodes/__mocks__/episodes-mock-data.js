import { PAGE_INFO } from '../../../components/page-layout/__mocks__/table-data';

const EPISODE_TABLE_MOCKED_DATA = {
  characters: [{ name: 'Tom' }],
  episode: '0E12',
  id: 1,
  name: 'Sunny day',
};

export const EPISODES_MOCKED_DATA = {
  episodes: {
    results: [EPISODE_TABLE_MOCKED_DATA],
    info: PAGE_INFO,
  },
};
