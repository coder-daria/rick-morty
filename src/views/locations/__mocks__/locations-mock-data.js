import { PAGE_INFO } from '../../../components/page-layout/__mocks__/table-data';

const LOCATION_TABLE_MOCKED_DATA = {
  id: 1,
  created: '11/25/2019, 4:59:08 PM',
  dimension: '',
  name: 'Sunny day',
  type: '',
  residents: [{ name: 'Tom' }],
};

export const LOCATIONS_MOCKED_DATA = {
  locations: {
    results: [LOCATION_TABLE_MOCKED_DATA],
    info: PAGE_INFO,
  },
};
