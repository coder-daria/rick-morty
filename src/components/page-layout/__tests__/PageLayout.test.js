import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { TABLE_DATA } from '../__mocks__/table-data';

import PageLayout from '../PageLayout';

const setup = () => {
  // eslint-disable-next-line react/jsx-props-no-spreading
  return shallow(<PageLayout {...TABLE_DATA} />);
};

describe('<PageLayout />', () => {
  it('should match the snapshot', () => {
    const wrapper = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
