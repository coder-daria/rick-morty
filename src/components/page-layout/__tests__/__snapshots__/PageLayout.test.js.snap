// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageLayout /> should match the snapshot 1`] = `
<div>
  <styled.div>
    <span>
      Showing
      <styled.span>
        1
      </styled.span>
      of
      <styled.span>
        1
      </styled.span>
       entries
    </span>
  </styled.div>
  <Styled(Table)
    columns={
      Array [
        Object {
          "dataIndex": "name",
          "key": "name",
          "title": "Name",
        },
        Object {
          "dataIndex": "species",
          "key": "species",
          "title": "Species",
        },
        Object {
          "dataIndex": "origin",
          "key": "origin",
          "title": "Origin",
        },
        Object {
          "dataIndex": "location",
          "key": "location",
          "title": "Location",
        },
        Object {
          "key": "action",
          "render": [Function],
          "title": "Action",
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "id": 1,
          "location": Object {
            "name": "Earth",
          },
          "name": "Rick Sanchez",
          "origin": Object {
            "name": "Earth",
          },
          "species": "Alien",
        },
      ]
    }
    loading={false}
    onChange={[Function]}
    pagination={
      Object {
        "pageSize": 20,
        "position": Array [
          "topLeft",
        ],
        "simple": true,
        "total": 1,
      }
    }
    rowKey={[Function]}
    scroll={
      Object {
        "y": 700,
      }
    }
    tableLayout="fixed"
  />
</div>
`;
